### PR TITLE
Workaround crash in atexit on NonStop platforms

### LINKED ISSUE #17537: Test 30-test_evp_extra broken on master and 3.1 on NonStop platforms
This test is broken around commit 57645339ab. This does not impact 3.0.1 or earlier. The reported platforms are NonStop ia64 (J06.22) and x86 (L21.06) built with c99.

The test fails with SIGSEGV during `atexit()` processing.
```
   ok 48 - test_ecx_short_keys
../../util/wrap.pl ../../test/evp_extra_test -context => 139
not ok 2 - running evp_extra_test with a non-default library context

#   Failed test 'running evp_extra_test with a non-default library context'
#   at test/recipes/30-test_evp_extra.t line 22.
```
with a trace:
```
(xInspect 1,971):bt
#0  0xfffffffff3bad428 in __process_atexit_functions ()
#1  0xfffffffff3c07871 in CRE_TERMINATOR_ ()
#2  0xfffffffff3d6ac54 in exit ()
#3  0x700d41aa in _MAIN () at \BLPROD.$RTAL.T8432ACM.CMAIN64C:52
```

### PR DESCRIPTION
**Pull request by t8m**
We cannot dynamically load the legacy provider into an application
that is linked statically to libcrypto as this causes
a double loading of libcrypto (one static and one dynamic) and
on NonStop this leads to a segfault in atexit().

Fixes #17537

### CODE REVIEW

**Comment by t8m** [comment]
> This fix breaks another subtest:
> 
> ```
> ../../util/wrap.pl ../../test/evp_extra_test -context => 139
> not ok 2 - running evp_extra_test with a non-default library context
> ```

Actually this is just the same error as without the workaround, isn't it? So the workaround doesn't work?? But why...

**Comment by rsbeckerca** [comment]
> @rsbeckerca is the stack trace still the same with the workaround?
> 

Forgot how to debug this directly

**Comment by rsbeckerca** [comment]
Same failure regardless. This happens because the DLL unloads invoking `atexit()` and then the program unloads reexecuting `atexit()`.

**Comment by rsbeckerca** [comment]
> > This happens because the DLL unloads invoking `atexit()` and then the program unloads reexecuting `atexit()`.
> 
> But `OPENSSL_init_crypto(OPENSSL_INIT_NO_ATEXIT, NULL);` should prevent that, shouldn't it? Hmm... actually it probably prevents only the statically linked libcrypto to avoid the atexit() calls. However the dynamic libcrypto linked from the legacy provider is still calling atexit().
> 
> I'll try a different workaround.

The fundamental issue is that data in the DLL exists twice, once in the main program's heap and another in the DLL heap, so depending on anything in the DLL is not going to work. Any semaphores should be declared in the main program and then passed into the DLL by reference - then you have a common allocation.

**Comment by t8m** [comment]
@rsbeckerca could you please try the updated workaround?

**Comment by rsbeckerca** [comment]
```
All tests successful.
Files=1, Tests=3, 11 wallclock secs ( 0.08 usr  0.00 sys +  9.31 cusr  0.01 csys =  9.40 CPU)
Result: PASS
```
I take this as good news

File: test/evp_extra_test.c
Comment (mattcaswell) [inline test/evp_extra_test.c:4734]: What is the purpose of loading the legacy provider if we can just not bother to load it like this?

**Comment by t8m** [comment]
Ping for reviews

**Comment by rsbeckerca** [comment]
Under test right now

**Comment by rsbeckerca** [comment]
This fix, already in `openssl-3.1` works.

**Comment by rsbeckerca** [comment]
The fix seemed to work, but needs to be merged.

**Comment by rsbeckerca** [comment]
The fix appears to work. Hoping for the merge.

**Comment by t8m** [comment]
I've marked it with 3.1.0 milestone so it should be merged before 3.1.0 final.

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Ping for second review

**Comment by levitte** [comment]
I'm not at all happy with this sort of hack, but yeahok.  That reminded me to rebase #17483, so I guess thanks :wink:

**Comment by t8m** [comment]
Still needs second approval - unless @levitte's comment above counts as such.

**Review by paulidale — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by paulidale** [comment]
and merged.

**Pull request closed without merging**
